### ISSUE TITLE
--nvccli: consider using `nvidia-container-cli list` and mounting the libraries manually

### ISSUE THREAD

**Issue by SomeoneSerge** [apptainer/apptainer#1893]
### Version of Apptainer

What version of Apptainer (or Singularity) are you using? Run

```
❯ apptainer --version
apptainer version 1.2.5
```

### Expected behavior

`apptainer exec --nv --nvccli ...` working without it or its dependencies having setuid bits, including when used with read-only images

### Actual behavior

What actually happened? Why was it incorrect?

- `apptainer` asserting that the image be writable: https://github.com/apptainer/apptainer/blob/dbaf1afa0e153e056c32dad2640b4d367a53ff14/internal/pkg/util/gpu/nvidia.go#L95-L97;
- the previous item patched out, `nvidia-container-cli ... --user configure` failing when trying to `prctl`-set capabilities unavailable to unprivileged users (e.g. `CAP_SYS_ADMIN`);
- the previous item patched out, `nvidia-container-cli` abusing glibc internals (e.g. using `/etc/ld.so.cache` which is specific to glibc and isn't guaranteed to exist on systems using glibc either);
- the previous item patched out (so that `nvidia-container-cli list` correctly generates the list of relevant libraries), `nvidia-container-cli ... --user configure` not affecting the final container's environment (nothing is written to `.singularity.d/libs`, nor do I observe any additional mounts in the container)

There clearly are several interconnected issues at play. It would seem that at this point libnvidia-container isn't entirely ready to be used in unprivileged environments (both in terms of capabilities and the write access), which is why I think using `list` instead of `configure` might be a reasonable step - for now. Apptainer could use `nvidia-container-cli list` in the same manner it uses `nvliblist.conf`. This could also be just a temporary change, until libnvidia-container has polished the support for unprivileged environments

### Steps to reproduce this behavior

How can others reproduce this issue/problem?

I'm somewhat constrained in time, so the easiest repro I can offer right now is using Nix. You can, however, tell me which patches to add and which commands to run, and I'll respond with the outputs. Otherwise here's one of the tests I use (Locked URL: `github:SomeoneSerge/nixpkgs/ea3581301525a64d6b60227bb498256816273446`):

```
❯ APPTAINER_MESSAGELEVEL=100000 NVIDIA_VISIBLE_DEVICES=all nix run --arg config '{ allowUnfree = true; }' -I nixpkgs=flake:github:SomeoneSerge/nixpkgs/feat/nixos-apptainer-nosuid -f '<nixpkgs>' apptainer.gpuChecks.saxpy.unwrapped
...
VERBOSE [U=1001,P=1319708] addCwdMount()                 /home/ss/Sources/nixpkgs/.worktree/apptainer found within container
DEBUG   [U=1001,P=1319708] create()                      nvidia-container-cli
DEBUG   [U=1001,P=1319732] findOnPath()                  Found "nvidia-container-cli" at "/nix/store/j9raldl3nsv797ydmc7v8kxgf8vbkks3-libnvidia-container-1.14.3/bin/nvidia-container-cli"
DEBUG   [U=1001,P=1319732] findOnPath()                  Found "ldconfig" at "/run/current-system/sw/bin/ldconfig"
DEBUG   [U=1001,P=1319732] NVCLIConfigure()              nvidia-container-cli binary: "/nix/store/j9raldl3nsv797ydmc7v8kxgf8vbkks3-libnvidia-container-1.14.3/bin/nvidia-container-cli" args: ["--debug" "--user" "configure" "--no-cgroups" "--device=all" "--compute" "--utility" "--ldconfig=@/run/current-system/sw/bin/ldconfig" "/nix/store/v1mwsgk404dryqfbffyafybyab1snzfp-apptainer-1.2.5/var/lib/apptainer/mnt/session/final"]
DEBUG   [U=1001,P=1319732] NVCLIConfigure()              Running nvidia-container-cli in user namespace
DEBUG   [U=1001,P=1319708] create()                      Chroot into /nix/store/v1mwsgk404dryqfbffyafybyab1snzfp-apptainer-1.2.5/var/lib/apptainer/mnt/session/final
DEBUG   [U=1001,P=1319732] Chroot()                      Hold reference to host / directory
DEBUG   [U=1001,P=1319732] Chroot()                      Called pivot_root on /nix/store/v1mwsgk404dryqfbffyafybyab1snzfp-apptainer-1.2.5/var/lib/apptainer/mnt/session/final
DEBUG   [U=1001,P=1319732] Chroot()                      Change current directory to host / directory
...
CUDA error at cudaRuntimeGetVersion(&rtVersion): CUDA driver version is insufficient for CUDA runtime version
```

### What OS/distro are you running

```sh
❯ cat /etc/os-release 
BUG_REPORT_URL="https://github.com/NixOS/nixpkgs/issues"
BUILD_ID="24.05.20231224.5f64a12"
DOCUMENTATION_URL="https://nixos.org/learn.html"
HOME_URL="https://nixos.org/"
ID=nixos
LOGO="nix-snowflake"
NAME=NixOS
PRETTY_NAME="NixOS 24.05 (Uakari)"
SUPPORT_URL="https://nixos.org/community.html"
VERSION="24.05 (Uakari)"
VERSION_CODENAME=uakari
VERSION_ID="24.05"
❯ nvidia-container-cli list
...
/nix/store/9x68sk5a6xc86q7fc6w8n40izdpwgv1q-nvidia-x11-545.29.06-6.1.69-lib32/lib/libcuda.so.545.29.06
...
❯ nvidia-container-cli info
NVRM version:   545.29.06
CUDA version:   12.3

Device Index:   0
Device Minor:   0
Model:          NVIDIA GeForce RTX 3090
...
```

### How did you install Apptainer

Write here how you installed Apptainer (or Singularity). Eg. RPM, source.

`nix build github:SomeoneSerge/nixpkgs/feat/nixos-apptainer-nosuid#apptainer` (the respective draft PR: https://github.com/NixOS/nixpkgs/pull/279235)

**Comment by DrDaveD** [comment]
We imported nvidia-container-cli support from [singularity-ce](https://github.com/sylabs/singularity) and so don't really know a lot about it.  We did not trust executing this external program from setuid-root mode so we didn't allow that.  As I understand it, it was designed for docker-like environments so it assumes it has the ability to create its own mountpoints, hence the need for the container to be writable.  

Let's back up.  I don't understand the reasons why someone would want to use nvidia-container-cli instead of `--nv`.  Why is it that you're wanting to use it?

I'm guessing that /etc/apptainer/nvliblist.conf with `--nv` does not pull in all the right libraries in your environment.  Is that correct?  What's missing?  I do see that running `nvidia-container-cli list` on a machine that has an nvidia gpu prints out a list of paths that look like things that might be needed to bind-mount into a container.  That's good to know.  So are you suggesting an option that does the equivalent of
```
-B `nvidia-container-cli list|paste -sd,`
```
?  I tried that on a CentOS7 system with an nvidia gpu that I have unprivileged access on, and then running `nvidia-smi` gave an error
```
NVIDIA-SMI couldn't find libnvidia-ml.so library in your system. Please make sure that the NVIDIA Display Driver is properly installed and present in your system.
Please also try adding directory that contains libnvidia-ml.so to your system PATH.
```
That's because the `nvidia-container-cli list` output includes only `/usr/lib64/libnvidia-ml.so.535.104.05` and not `/usr/lib64/libnvidia-ml.so`.

**Comment by SomeoneSerge** [comment]
> Let's back up.

👍🏻 I almost certainly chose the wrong starting point for the issue because I'm not too familiar with the project.

> I don't understand the reasons why someone would want to use `nvidia-container-cli` instead of `--nv`

Partly, by mistake: I didn't notice a local patch setting `use nvidia-container-cli = yes` and so thought that `--nvccli` was the default. That said, I meant for this issue to track `--nvccli` specifically (I opened https://github.com/apptainer/apptainer/issues/1894 for `nvliblist.conf`), since the option does exist.

Seeing how apptainer is generally unprivileged, I think it might be desirable to use `nvidia-container-cli` in a way that doesn't require anything more if that's possible.

> We imported nvidia-container-cli support from [singularity-ce](https://github.com/sylabs/singularity) and so don't really know a lot about it

Perhaps I should move this issue to singularity-ce then, and focus on nvliblist.conf with apptainer?

> As I understand it, it was designed for docker-like environments so it assumes it has the ability to create its own mountpoints, hence the need for the container to be writable.

It definitely does seem like they assume some privileges, but I'd wait for the official response. I'm not sure if the write access is part of these privileges, because that's not the failure I observe when I remove the early exit. Instead, I first encounter the issues setting the capabilities

> That's because the nvidia-container-cli list output includes only /usr/lib64/libnvidia-ml.so.535.104.05 and not /usr/lib64/libnvidia-ml.so

I see, I didn't notice that. I guess the soname symlinks could be created in an extra pass _if_ we were to use the `list` approach

**Comment by DrDaveD** [comment]
Yes if you want to continue working with nvidia-container-cli and can reproduce the issue with singularity-ce it would be better to start there.

**Comment by SomeoneSerge** [comment]
I'd like to keep this issue open until there's a response from nvidia - e.g. in case we could find a better error message than "usr/bin isn't writable"
